File: app/models/location.rb
class Location < ActiveRecord::Base
  belongs_to :car
  belongs_to :origin, foreign_key: "origin_id", class_name: "City"
  belongs_to :destination, foreign_key: "destination_id", class_name: "City"
  belongs_to :current, foreign_key: "current_id", class_name: "City"

  delegate :state, to: :origin, prefix: :origin, allow_nil: true
  delegate :state, to: :destination, prefix: :destination, allow_nil: true
  delegate :state, to: :current, prefix: :current, allow_nil: true

  def full(method)
    raise MethodNotFound unless [:origin, :destination, :current].include? method
    city = self.send method
    "#{city.name.titleize}, #{city.state.acronym}" if city
  end

  def as_json(options={})
    super(
      include: {
        origin_state: { only: [:id] },
        destination_state: { only: [:id] },
        current_state: { only: [:id] }
      }
    )
  end
end

File: app/assets/javascripts/angular/models/location.js
(function(angular) {
  "use strict";

  angular.module("models.location", [
    "models.location"
    ])
  .factory("Location", function() {
    function Location(params) {
      if (!params) params = {};

      this.origin = {id: params.originId};
      this.destination = {id: params.destinationId};
      this.current = {id: params.currentId};
      this.originState = {id: params.originStateId};
      this.destinationState = {id: params.destinationStateId};
      this.currentState = {id: params.currentStateId};
    }

    Location.fromJson = function(json) {
      return new Location({
        originId: json.origin_id,
        destinationId: json.destination_id,
        currentId: json.current_id,
        originStateId: json.origin_state ? json.origin_state.id : null,
        destinationStateId: json.destination_state ? json.destination_state.id : null,
        currentStateId: json.current_state ? json.current_state.id : null
      });
    }

    return Location;
  });
})(window.angular);
